### PR TITLE
tidb_query: fix panic when input of cast_string_as_time is invalid UTF-8 bytes

### LINKED ISSUE #9970: index out of bounds: the len is 4 but the index is 4
### What version of TiKV are you using?

```bash
[tidb@vm4 bin]$ ./tikv-server -V
TiKV
Release Version:   5.0.0-rc.x
Edition:           Community
Git Commit Hash:   4807a3d610f2809bd6cc5b1e7645dbb01d5e9e52
Git Commit Branch: heads/refs/tags/v5.0.0-nightly
UTC Build Time:    2021-04-02 16:35:25
Rust Version:      rustc 1.51.0-nightly (bc39d4d9c 2021-01-15)
Enable Features:   jemalloc mem-profiling portable sse protobuf-codec test-engines-rocksdb cloud-aws cloud-gcp
Profile:           dist_release
```
### Steps to reproduce

sorry, I can't find out the SQL which leads to this error occurs. The stack logs may be useful to find it out.
### What did you expect?
tikv doesn't exit and  restart 
### What did happened?
tikv exited with fatal error and  restart 
```bash
[2021/04/06 20:24:41.775 +08:00] [FATAL] [lib.rs:465] ["index out of bounds: the len is 4 but the index is 4"] [backtrace="stack backtrace:\n   0: tikv_util::set_panic_hook::{{closure}}\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tikv_util/src/lib.rs:464\n   1: std::panicking::rust_panic_with_hook\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35//library/std/src/panicking.rs:595\n   2: std::panicking::begin_panic_handler::{{closure}}\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35//library/std/src/panicking.rs:497\n   3: std::sys_common::backtrace::__rust_end_short_backtrace\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35//library/std/src/sys_common/backtrace.rs:141\n   4: rust_begin_unwind\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35//library/std/src/panicking.rs:493\n   5: core::panicking::panic_fmt\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35//library/core/src/panicking.rs:92\n   6: core::panicking::panic_bounds_check\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35//library/core/src/panicking.rs:69\n   7: core::unicode::unicode_data::skip_search\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35//library/core/src/unicode/unicode_data.rs:75\n      core::unicode::unicode_data::white_space::lookup\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35//library/core/src/unicode/unicode_data.rs:541\n   8: core::char::methods::<impl char>::is_whitespace\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/core/src/char/methods.rs:794\n      core::str::<impl str>::trim::{{closure}}\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/core/src/str/mod.rs:1727\n      <F as core::str::pattern::MultiCharEq>::matches\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/core/src/str/pattern.rs:600\n      <core::str::pattern::MultiCharEqSearcher<C> as core::str::pattern::Searcher>::next\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/core/src/str/pattern.rs:644\n      core::str::pattern::Searcher::next_reject\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/core/src/str/pattern.rs:242\n      <core::str::pattern::CharPredicateSearcher<F> as core::str::pattern::Searcher>::next_reject\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/core/src/str/pattern.rs:736\n      core::str::<impl str>::trim_matches\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/core/src/str/mod.rs:1926\n      core::str::<impl str>::trim\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/core/src/str/mod.rs:1727\n   9: tidb_query_datatype::codec::mysql::time::parser::parse\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_datatype/src/codec/mysql/time/mod.rs:565\n      tidb_query_datatype::codec::mysql::time::Time::parse\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_datatype/src/codec/mysql/time/mod.rs:738\n  10: tidb_query_expr::impl_cast::cast_string_as_time\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_expr/src/impl_cast.rs:1148\n  11: <tidb_query_expr::types::function::Arg<Arg0_,tidb_query_expr::types::function::Null> as tidb_query_expr::impl_cast::CastStringAsTime_Fn>::eval\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_expr/src/impl_cast.rs:1140\n      <tidb_query_expr::impl_cast::CastStringAsTime_Evaluator as tidb_query_expr::types::function::Evaluator>::eval\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_expr/src/impl_cast.rs:1140\n      <tidb_query_expr::types::function::ArgConstructor<A,E> as tidb_query_expr::types::function::Evaluator>::eval\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_expr/src/types/function.rs:259\n      tidb_query_expr::impl_cast::cast_string_as_time_fn_meta::run\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_expr/src/impl_cast.rs:1140\n  12: tidb_query_expr::types::expr_eval::<impl tidb_query_expr::types::expr::RpnExpression>::eval_decoded\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_expr/src/types/expr_eval.rs:248\n  13: <tidb_query_executors::fast_hash_aggr_executor::FastHashAggregationImpl as tidb_query_executors::util::aggr_executor::AggregationExecutorImpl<Src>>::process_batch_input\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_expr/src/types/expr_eval.rs:161\n  14: tidb_query_executors::util::aggr_executor::AggregationExecutor<Src,I>::handle_next_batch\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_executors/src/util/aggr_executor.rs:205\n      <tidb_query_executors::util::aggr_executor::AggregationExecutor<Src,I> as tidb_query_executors::interface::BatchExecutor>::next_batch\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_executors/src/util/aggr_executor.rs:282\n      <tidb_query_executors::fast_hash_aggr_executor::BatchFastHashAggregationExecutor<Src> as tidb_query_executors::interface::BatchExecutor>::next_batch\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_executors/src/fast_hash_aggr_executor.rs:49\n      <tidb_query_common::execute_stats::WithSummaryCollector<C,T> as tidb_query_executors::interface::BatchExecutor>::next_batch\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_executors/src/interface.rs:108\n  15: <alloc::boxed::Box<T> as tidb_query_executors::interface::BatchExecutor>::next_batch\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_executors/src/interface.rs:77\n      tidb_query_executors::runner::BatchExecutorsRunner<SS>::internal_handle_request\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_executors/src/runner.rs:529\n  16: tidb_query_executors::runner::BatchExecutorsRunner<SS>::handle_request::{{closure}}\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tidb_query_executors/src/runner.rs:414\n      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/core/src/future/mod.rs:80\n      <tikv::coprocessor::dag::BatchDAGHandler as tikv::coprocessor::RequestHandler>::handle_request::__handle_request::{{closure}}\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/src/coprocessor/dag/mod.rs:105\n      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/core/src/future/mod.rs:80\n      <core::pin::Pin<P> as core::future::future::Future>::poll\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/core/src/future/future.rs:119\n      <minitrace::future::TraceWrapped<T> as core::future::future::Future>::poll\n             at /rust/git/checkouts/minitrace-rust-4a91d55623f07cfc/de69110/src/future.rs:107\n[2021/04/06 20:24:57.103 +08:00] [INFO] [lib.rs:89] ["Welcome to TiKV"]
```

### PR DESCRIPTION
**Pull request by sticnarf**
### What problem does this PR solve?

Issue Number: close #9970

### What is changed and how it works?

The underlying bytes of a string may not be valid UTF-8 bytes. It is unsafe to convert it to a `str` without checking. In #9970, it causes a panic when calling `trim` on the `str`.

### Check List 

Tests 

- Unit test

### Release note 

- Fix potential panics when input of cast_string_as_time is invalid UTF-8 bytes

### CODE REVIEW

**Review by skyzh — APPROVED** [review]

**Comment by sticnarf** [comment]
@skyzh I realize MySQL actually returns warnings. So I change to use `String::from_utf8_lossy` instead. (It seems you approved before I made the second change)

**Comment by skyzh** [comment]
> @skyzh I realize MySQL actually returns warnings. So I change to use `String::from_utf8_lossy` instead. (It seems you approved before I made the second change)

Then the function won't return error, and it seems that test cases need to be modified.

File: components/tidb_query_expr/src/impl_cast.rs [hunk 1]
```diff
@@ -1145,9 +1145,10 @@ fn cast_string_as_time(
 ) -> Result<Option<Time>> {
     if let Some(val) = val {
         // Convert `val` to a string first and then parse it as a float string.
+        let val = String::from_utf8_lossy(val);
```
Comment (iosmanthus) [inline components/tidb_query_expr/src/impl_cast.rs:1148]: I am a little confused here, where will the expected warning will produced.

Comment (sticnarf) [reply components/tidb_query_expr/src/impl_cast.rs:1148]: If it is not valid UTF-8 bytes, the converted string will contain a replacement character `U+FFFD`. This character cannot be parsed so the parser will generate the warning.

Comment (iosmanthus) [reply components/tidb_query_expr/src/impl_cast.rs:1148]: I got it!

**Review by iosmanthus — APPROVED** [review]
LGTM

**Pull request merged**